File: hackerrank/problem solving/algorithms/2 - implementation/41 - taum and bday/taum_and_bday.go
// https://www.hackerrank.com/challenges/taum-and-bday/problem?isFullScreen=true

package main

import "fmt"

type Gifts struct {
    nBlackGifts               int
    nWhiteGifts               int
    blackGiftCost             int
    whiteGiftCost             int
    costToConvertBetweenGifts int
}

func main() {
    var n int
    _, _ = fmt.Scan(&n)
    for range n {
        fmt.Println(minCostOfBuyingGifts(readTestCase()))
    }
}

func readTestCase() Gifts {
    var x Gifts
    _, _ = fmt.Scan(&x.nBlackGifts, &x.nWhiteGifts, &x.blackGiftCost, &x.whiteGiftCost, &x.costToConvertBetweenGifts)
    return x
}

func minCostOfBuyingGifts(x Gifts) int {
    blackCost := min(x.blackGiftCost, x.whiteGiftCost+x.costToConvertBetweenGifts)
    whiteCost := min(x.whiteGiftCost, x.blackGiftCost+x.costToConvertBetweenGifts)
    return x.nBlackGifts*blackCost + x.nWhiteGifts*whiteCost
}

File: hackerrank/problem solving/algorithms/2 - implementation/41 - taum and bday/taum_and_bday.py
# https://www.hackerrank.com/challenges/taum-and-bday/problem?isFullScreen=true

from dataclasses import dataclass


@dataclass(frozen=True)
class Gifts:
    n_black_gifts: int
    n_white_gifts: int
    black_gift_cost: int
    white_gift_cost: int
    cost_to_convert_between_gifts: int


def main():
    n = int(input())
    for _ in range(n):
        print(min_cost_of_buying_gifts(read_test_case()))


def read_test_case():
    arr = read_numbers()
    arr.extend(read_numbers())
    return Gifts(*arr)


def read_numbers():
    return list(map(int, input().split()))


# T: O(1)
# S: O(1) extra space
def min_cost_of_buying_gifts(x):
    black_cost = min(x.black_gift_cost, x.white_gift_cost + x.cost_to_convert_between_gifts)
    white_cost = min(x.white_gift_cost, x.black_gift_cost + x.cost_to_convert_between_gifts)
    return x.n_black_gifts * black_cost + x.n_white_gifts * white_cost


if __name__ == '__main__':
    main()
